llama.cpp

```diff
@@ -740,32 +747,55 @@ struct llama_model_loader {
 
         if (use_mmap) {
             mapping.reset(new llama_mmap(&file_loaders.at(0)->file, prefetch_size));
-            if (!lmlock) {
-                // Don't call the callback since the actual loading will be lazy
-                // and we can't measure it.
-                progress_callback = NULL;
-            }
             if (lmlock) {
                 lmlock->init(mapping->addr);
             }
         }
 
         size_t done_size = 0;
         for (llama_load_tensor & lt : tensors_map.tensors) {
-            if (lt.ggml_tensor->backend != GGML_BACKEND_CPU) {
-                continue;
-            }
             if (progress_callback) {
                 progress_callback((float) done_size / data_size, progress_callback_user_data);
             }
             LLAMA_ASSERT(lt.ggml_tensor); // unused tensors should have been caught by load_data already
             lt.data = (uint8_t *) lt.ggml_tensor->data;
+
+            // allocate temp buffer if not using mmap
+            if (!use_mmap && lt.data == NULL) {
+                lt.data = (uint8_t*)malloc(ggml_nbytes(lt.ggml_tensor));
```

Comment:
How does this get freed again for `GGML_BACKEND_CPU`?

Reply:
no. The memory will be used as tensors' data.

Reply:
What I mean is this: when you use llama.cpp from the command line the memory automatically gets freed by the operating system once the process exits. However, when you use the Python bindings you can create a llama.cpp model object and then delete it again. How would the memory get freed in that scenario?

Reply:
I didn't change the logic of deallocate tensor here. as the buffer is kept in tensor.data, it will get freed when tensor is deleted. (in ~llama_model)

Reply:
Sorry, but I really can't follow your logic here. I can't see any use of `malloc` in this part of the code on master. Therefore, if this indeed is a new memory allocation then it needs to be freed somewhere. And I don't see it being freed in `~llama_model` either. That destructor frees the memory buffer allocated by the `ggml_context` but the memory buffer allocated here is not part of the buffer. The method `ggml_cuda_free_data` doesn't do it either.

And just in case there is a general misconception: `ggml_tensor` holds a *pointer* to data, not the data itself. So deleting a `ggml_tensor` does **not** result in the data being freed.

Reply:
line 765 will never hit if backend is CPU. What I mean the memory allocate/deallocate is really refering VRAM instead of RAM.

Reply:
It seems you're right. I misunderstood how the code works. I suggest we add an assert to make the way the code works easier to understand:

```suggestion
                GGML_ASSERT(lt.ggml_tensor->backend != GGML_BACKEND_CPU);
                lt.data = (uint8_t*)malloc(ggml_nbytes(lt.ggml_tensor));
```